parasol/indexing.py

```diff
@@ -0,0 +1,145 @@
+"""
+Model-based indexing with Solr.
+
+Items to be indexed in Solr should extend :class:`Indexable`. The
+default implementation should work for most Django models; at a minimum
+you should extend :meth:`Indexable.index_data` to include the information
+to be indexed in Solr. You may also customize :meth:`Indexable.index_item_type`
+and :meth:`Indexable.index_item_id`.
+
+To manually index content in Solr, see
+:mod:`~parasol.management.commands.index` manage command documentation.
+
+-------------------------
+
+"""
+
+import itertools
+import logging
+
+try:
+    from django.db.models.query import QuerySet
+
+    from parasol.solr.django import SolrClient
+except ImportError:
+    QuerySet = SolrClient = None
+
+
+logger = logging.getLogger(__name__)
+
+
+class Indexable:
+    """Mixin for objects that are indexed in Solr.  Subclasses must implement
+    `index_id` and `index` methods.
+    """
+
+    # NOTE: current implementation is Django-specific, intended for
+    # use with django models. Should be possible to generalize once
+    # we have other use cases.
+
+    #: number of items to index at once when indexing a large number of items
+    index_chunk_size = 150
+
+    #: solr connection
+    solr = None
+
+    #: id separator for auto-generated index ids
+    ID_SEPARATOR = '.'
+
+    def __init__(self):
+        # initialize connection to solr on first instance initialization
+        Indexable._init_solr()
+
+    @classmethod
+    def _init_solr(cls):
+        # store on the class to take advantage of sessions
+        if cls.solr is None:
+            cls.solr = SolrClient()
+
+    @classmethod
+    def all_indexables(cls):
+        """Find all :class:`Indexable` subclasses for indexing."""
+        return cls.__subclasses__()
+
+    @classmethod
+    def index_item_type(cls):
+        """Label for this kind of indexable item. Must be unique
+        across all Indexable items in an application. By default, uses
+        Django model verbose name. Used in default index id and
+        in index manage command. """
+        return cls._meta.verbose_name
+
+    def index_id(self):
+        """Solr identifier. By default, combines :meth:`index item_type`
+        and :attr:`id` with :attr:ID_SEPARATOR`."""
+        return '{}{}{}'.format(self.index_item_type(), self.ID_SEPARATOR,
```

Comment:
Which do you think is 'better' style at this point, the old `printf` style interpolation or format? I've gone back and forth and don't have strong feelings -- at this point in the PEP process, I don't think `printf` style is in danger of going away any time soon, while printf is very elegant for passing a dictionary and using its keys.

Reply:
My understanding is that `format` is considered superior but I'm more familiar with the % notation and tend to use that by default. I _thought_ I'd read that `format` was faster than `%`, but I'm looking again and seeing the reverse?
The number formatting seems better with `format` for some uses, e.g. trivially add commas for thousands separators. 

Actually, if we care about speed I should probably switch to `''.join()` here...

The new f-string syntax looks interesting, but not sure I want to get used to a something that requires python 3.6.

Do you think we should try to be consistent about using one or the other?

Reply:
I think we should stick to what makes sense in the context but realized we'd never had the conversation. I prefer the `printf` (and I think it helps make the code more familiar to those coming to it not from a python background, imo -- because printf styling is pretty close to a lingua franca).

I'd say, use `format` when it's advantageous for functionality or clarity, and otherwise `printf`.